### DISCUSSION TITLE
How to reexport a submodule from the root?

### DISCUSSION THREAD

**Discussion by rchiodo** [microsoft/pyright#8162] [Q&A]
Suppose I had a package defined like so:

```
mypackage
   - __init__.py 
   childpackage
       - __init__.py
       grandchildpackage
           - foo.py
```

And I wanted a user to be able to do something like so:

``` python
from mypackage.grandchildpackage import Foo
```

Now there is a way to get this to work in the runtime (as explained [here](https://dev.to/bastantoine/playing-with-python-path-attribute-3ic3)). I add the `childpackage` path to `__path__` in the `__init__.py`. Like so:

```
__path__.append(os.path.join(os.path.dirname(__file__), "childpackage"))
```

But is there a way to model this behavior so that Pyright can understand it?

How do I eliminate this error:

![image](https://github.com/microsoft/pyright/assets/19672699/9f5edf28-44fe-41a7-a385-23e82340d87d)

I bring this up because of this issue in Keras:
https://github.com/keras-team/keras/issues/19779

**Comment by rchiodo** [comment]
My initial though was that this would work in the `mypackage/__init__.py`

```python
from .childpackage import grandchildpackage as grandchildpackage
```

But that only seems to work for symbols, not modules.

**Reply by heejaechang** [reply]
can pyright add heuristics to support this pattern? https://dev.to/bastantoine/playing-with-python-path-attribute-3ic3

it seems there are a few well known packages using this trick.

**Reply by erictraut** [reply]
Such heuristics would be incredibly hacky and difficult to debug when they fail. In the past, we've always said that we don't support dynamic manipulation of `__path__`, and that's the stance I'd like to maintain.

**Reply by rchiodo** [reply]
Do you think this would be possible though?

```
from .childpackage import grandchildpackage as grandchildpackage
```

Reexporting a module from an `__init__.py` which effectively changes the path we look for the grandchildpackage in?

**Reply by rchiodo** [reply]
Oh I guess that doesn't work at runtime by itself though. It still requires the dynamic path addition.

**Reply by erictraut** [reply]
That sounds potentially feasible. It's a pattern that I've never seen used before, which explains why it was never implemented.

This would involve touching code that is not currently fresh in my memory, so I (or someone else) would need to investigate what would be involved — and whether it would break any assumptions in the current code or introduce significant performance issues with import resolution.

**Reply by heejaechang** [reply]
> Such heuristics would be incredibly hacky and difficult to debug when they fail. In the past, we've always said that we don't support dynamic manipulation of __path__, and that's the stance I'd like to maintain.

we might not support every possible ways but like `__all__`, we could handle a few specific case or pattern? then we can tell package maintainer that if you use this pattern, `pyright` will understand it and will work at runtime as well. not ideal but still a compromise?

**Reply by erictraut** [reply]
I don't think we should entertain supporting dynamic manipulation of `__paths__`. Please don't pursue that Heejae.

**Reply by erictraut** [reply]
In the case of `__all__`, we eventually got to the point where we standardized a limited subset of forms that all tools can and should support. I don't think we should attempt a similar standardization effort for dynamic `__paths__` manipulation. Other tools like mypy likewise do not handle this, and for good reason.
